### PR TITLE
[iOS][expo-constants] migrate expo-contants to new modules API

### PR DESCRIPTION
**Pull request by alanjhughes**
# Why
Continue migration of expo modules to the new API

# How
Followed the usual steps involved in migrating a module

# Test Plan
All original tests pass as before and confirmed in the bare-expo app

### CODE REVIEW

File: apps/bare-expo/ios/Podfile.lock [hunk 1]
```diff
@@ -1,8 +1,8 @@
 PODS:
   - boost (1.76.0)
-  - CocoaLumberjack (3.7.4):
-    - CocoaLumberjack/Core (= 3.7.4)
-  - CocoaLumberjack/Core (3.7.4)
+  - CocoaLumberjack (3.8.0):
+    - CocoaLumberjack/Core (= 3.8.0)
+  - CocoaLumberjack/Core (3.8.0)
```
Comment (tsapeta) [inline apps/bare-expo/ios/Podfile.lock:5]: Could you revert changes in the lock and rerun `pod install`? It's better not to upgrade unrelated dependencies 😉

File: packages/expo-constants/CHANGELOG.md [hunk 2]
```diff
@@ -6,6 +6,8 @@
 
 ### 🎉 New features
 
+Migrated iOS codebase to use the new Expo modules API. ([#20002](https://github.com/expo/expo/pull/20002) by [@alanhughes](https://github.com/alanjhughes))
```
Comment (tsapeta) [inline packages/expo-constants/CHANGELOG.md:9]:
```suggestion
- Migrated iOS codebase to use the new Expo modules API. ([#20002](https://github.com/expo/expo/pull/20002) by [@alanhughes](https://github.com/alanjhughes))
```

File: packages/expo-constants/ios/ConstantsModule.swift [hunk 3]
```diff
@@ -0,0 +1,36 @@
+//
+//  ConstantsModule.swift
+//  ExpoConstants-ExpoConstants
+//
+//  Created by Alan Hughes on 11/11/2022.
+//
+
```
Comment (tsapeta) [inline packages/expo-constants/ios/ConstantsModule.swift:7]:
```suggestion
```

File: packages/expo-constants/ios/ConstantsModule.swift [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    private lazy var constants = {
+        let dict = appContext?.constants?.constants()
+
+        if let result = dict as? [String: Any?] {
+            return result
+        }
+
+        return [String: Any?]()
+    }()
```
Comment (tsapeta) [inline packages/expo-constants/ios/ConstantsModule.swift:21]: Probably it can be simplified to:

```suggestion
    private lazy var constants = appContext?.constants?.constants() as? [String: Any] ?? [:]
```

File: packages/expo-constants/ios/ConstantsModule.swift [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+        Constants {
+            return constants
+        }
+
+        AsyncFunction("getWebViewUserAgentAsync") { (promise: Promise) in
+            let webView = WKWebView()
+            let result = webView.value(forKey: "userAgent")
+            promise.resolve(result)
+        }.runOnQueue(.main)
```
Comment (tsapeta) [inline packages/expo-constants/ios/ConstantsModule.swift:34]:
```suggestion
        AsyncFunction("getWebViewUserAgentAsync") {
            let webView = WKWebView()
            return webView.value(forKey: "userAgent")
        }.runOnQueue(.main)
```

File: packages/expo-constants/ios/EXConstants.podspec [hunk 6]
```diff
@@ -16,11 +16,17 @@ Pod::Spec.new do |s|
 
   s.dependency 'ExpoModulesCore'
 
+  # Swift/Objective-C compatibility	
+  s.pod_target_xcconfig = {	
+    'DEFINES_MODULE' => 'YES',	
+    'SWIFT_COMPILATION_MODE' => 'wholemodule'	
+  }
```
Comment (tsapeta) [inline packages/expo-constants/ios/EXConstants.podspec:23]: Can you also change the pod name to `ExpoConstants`?

File: packages/expo-constants/ios/EXConstantsInstallationIdProvider.m [hunk 7]
```diff
@@ -1,6 +1,6 @@
 // Copyright 2015-present 650 Industries. All rights reserved.
 
-#import <EXConstants/EXConstantsInstallationIdProvider.h>
+#import <EXConstantsInstallationIdProvider.h>
```
Comment (tsapeta) [inline packages/expo-constants/ios/EXConstantsInstallationIdProvider.m:3]: Is this change necessary? 🤔

**Review by tsapeta — CHANGES_REQUESTED** [review]
Nice! Could you also reindent Swift files? We prefer using 2 spaces 😉

File: packages/expo-constants/ios/ConstantsModule.swift (hunk 5 above)
Comment (alanjhughes) [reply packages/expo-constants/ios/ConstantsModule.swift:34]: This results in a compilation error `Generic parameter 'R' could not be inferred`. value(forKey:) is returning Any?. Casting doesn't seem to help. 

AsyncFunction("getWebViewUserAgentAsync") { () -> Any? in
How about this?

File: packages/expo-constants/ios/EXConstants.podspec (hunk 6 above)
Comment (alanjhughes) [reply packages/expo-constants/ios/EXConstants.podspec:23]: Will changing it not require a lot of changes in other modules?

File: packages/expo-constants/ios/ConstantsModule.swift (hunk 5 above)
Comment (tsapeta) [reply packages/expo-constants/ios/ConstantsModule.swift:34]: Hmm, I think `return webView.value(forKey: "userAgent") as? String` should work 🤔

File: packages/expo-constants/ios/EXConstants.podspec (hunk 6 above)
Comment (tsapeta) [reply packages/expo-constants/ios/EXConstants.podspec:23]: Right... it will 😞 Let's leave it as is for now then. I'll take care of updating everything later 😉

File: packages/expo-constants/ios/ConstantsModule.swift (hunk 5 above)
Comment (alanjhughes) [reply packages/expo-constants/ios/ConstantsModule.swift:34]: No, that won't work either. I don't like the Any? either but not sure what's best to

Comment (tsapeta) [reply packages/expo-constants/ios/ConstantsModule.swift:34]: Interesting, looks like the compiler is not smart enough and we need to tell him explicitly what's the return type 😅 
Try this then:
```swift
AsyncFunction("getWebViewUserAgentAsync") { () -> String? in
  let webView = WKWebView()
  return webView.value(forKey: "userAgent") as? String
}.runOnQueue(.main)
```

Comment (alanjhughes) [reply packages/expo-constants/ios/ConstantsModule.swift:34]: Yes, that'll work 👍

**Review by tsapeta — APPROVED** [review]
Looks good now, thank you! 🚀
